From bb6f16c237c11a3733b11d72db2aa2438718d18e Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Sun, 7 Sep 2025 17:58:03 +0200
Subject: [PATCH 21/22] board: marvell: octeontex2_cn913x: service watchdog
 before starting linux

Service the watchdog immediately before starting linux, to ensure there
is enough time for userspace to start servicing watchdog on its own.

The time of starting linux is subject to user intervention and can
therefore happen shortly before watchdog expires, leading to unexpected
boot failures.
Further it is possible for u-boot initialisation and timeout together
cause similar issues.

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 board/Marvell/octeontx2_cn913x/board.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/board/Marvell/octeontx2_cn913x/board.c b/board/Marvell/octeontx2_cn913x/board.c
index 9e117436e5..abd0e3f7ae 100644
--- a/board/Marvell/octeontx2_cn913x/board.c
+++ b/board/Marvell/octeontx2_cn913x/board.c
@@ -348,3 +348,20 @@ int last_stage_init(void)
 	return 0;
 }
 #endif /* CONFIG_LAST_STAGE_INIT */
+
+void board_prep_linux(void *) {
+	struct uclass *uc;
+	static struct udevice *dev;
+	int ret;
+
+	ret = uclass_get(UCLASS_WDT, &uc);
+	if (ret)
+		 return;
+
+	/* reset any watchdog to give linux sufficient starting time */
+	uclass_foreach_dev(dev, uc) {
+		 ret = wdt_reset(dev);
+		 if (!ret)
+			  printf("serviced %s\n", dev->name);
+	}
+}
-- 
2.51.0

